manual/src/cmds/runtime-tracing.etex

```diff
@@ -0,0 +1,266 @@
+consume events or choose to only do so in response to some circumstance, e.g if
+a particular query or operation takes longer than expected to complete.
+
+There are three main classes of events emitted by runtime events:
+
+\begin{description}
+\item[Spans] Events having a duration in time. An example of a span event in the
+OCaml garbage collector would be a a minor collection. Spans may contain other
+spans, e.g there may be other spans that begin after a minor collection has
+begun and end before it does.
+\item[Lifecycle events] Events that occur at a moment in time. An example of a
+lifecycle event would be the termination of a domain.
+\item[Counters] Events that include the measurement of some quantity of
+interest. An example of a counter would be the number of words promoted from
+the minor to the major heap during the last minor garbage collection.
+\end{description}
+
+\section{s:runtime-tracing-architecture}{Architecture}
+
+The runtime tracing system can be split in to two parts. The probes which emit
+events and the rest of the system that ingest and transport them.
+
+\subsection{s:runtime-tracing-probes}{Probes}
+Probes collect performance events from the runtime system. These are further
+split in to two sets, those always available and those only available in the
+instrumented runtime. Probes in the instrumented runtime are primarily of
+interest to developers of the OCaml runtime and garbage collector and, at
+present, only consist of major heap allocation size counter events.
+
+The full set of events emitted by probes and their documentation can be found in
+\ifouthtml
+ \ahref{libref/Runtime_events.html}{Module \texttt{Runtime_events}}.
+\else
+ section~\ref{Runtime_events}.
+\fi
+
+\subsection{s:runtime-tracing-ingestion}{Events transport}
+
+The second part of the system ingests events emitted by the probes and makes
+these available to consumers.
+
+\subsubsection{s:runtime-tracing-ringbuffers}{Ring buffers}
+
+Events are transported using a data structure known as a {\em ring buffer}. This
+data structure consists of two pointers into a linear backing array, the tail
+pointer points to a location where new events can be written and the head
+pointer points to the oldest event in the buffer that can be read. When
+insufficient space is available in the backing array to write new events, the
+head pointer is advanced and the oldest events are overwritten by new ones.
+
+The ring buffer implementation used in runtime events can be written by at most
+one producer at a time but can be read simultaneously by multiple consumers
+without coordination from the producer. There is a unique ring buffer for every
+running domain and on domain termination ring buffers may be re-used for newly
+spawned domains. The ring buffers themselves are stored in a memory-mapped file
+with the processes identifier as the name and the extension ".events", this
+enables them to be read from outside the main OCaml process.
```

Comment:
Considering we promise that the ".events" files can be read from outside the main OCaml process, I think we should tell people what the file format is. That is, tell them what the file contains and link to the place in the code where people find the corresponding data structures.

Reply:
I think adding a link to the documentation `\libdocitem{Runtime_events}{runtime events}` (as done for systhreads in the `libthreads.etex` manual page.) should be good enough.
It would provide information on how to process these programmatically from an external program.
(As well as the basics outlined later in this chapter)